ui/seek_bar.js

```diff
@@ -677,6 +693,201 @@ shaka.ui.SeekBar = class extends shaka.ui.RangeElement {
       return minutes + ':' + seconds;
     }
   }
+
+  /**
+   * Sets up the chapter element creator and change handling.
+   * @private
+   */
+  setupChapters_() {
+    let language = 'und';
+    /** @type {!Array<shaka.extern.Chapter>} */
+    let chapters = [];
+
+    /**
+    * Does a value compare on chapters.
+    * @param {shaka.extern.Chapter} a
+    * @param {shaka.extern.Chapter} b
+    * @return {boolean}
+    */
+    const chaptersEqual = (a, b) => {
+      return (!a && !b) || (a.id === b.id && a.title === b.title &&
+          a.startTime === b.startTime && a.endTime === b.endTime);
+    };
+
+    /** @type {function(): void} */
+    const handleChapterTrackChange = () => {
+      let nextLanguage = 'und';
+      /** @type {!Array<shaka.extern.Chapter>} */
+      let nextChapters = [];
+
+      const currentLocales = this.localization.getCurrentLocales();
+      for (const locale of Array.from(currentLocales)) {
+        nextLanguage = locale;
+        nextChapters = this.player.getChapters(nextLanguage);
+        if (nextChapters.length) {
+          break;
+        }
+      }
+      if (!nextChapters.length) {
+        nextLanguage = 'und';
+        nextChapters = this.player.getChapters(nextLanguage);
+      }
+
+      const languageChanged = nextLanguage !== language;
+      const chaptersChanged = chapters.length !== nextChapters.length ||
+        !chapters.some((c, idx) => {
+          const n = nextChapters.at(idx);
+          return chaptersEqual(c, n) ||
+            nextChapters.some((n) => chaptersEqual(c, n));
+        });
+
+      language = nextLanguage;
+      chapters = nextChapters;
+      if (!nextChapters.length) {
+        this.deletePreviousChapters_();
+      } else if (languageChanged || chaptersChanged) {
+        this.createChapterElements_(this.container, chapters);
+      }
+    };
+
+    handleChapterTrackChange();
+
+    this.eventManager.listen(
+        this.player, 'unloading', () => {
+          this.deletePreviousChapters_();
```

Comment:
Hm. I'm trying to think this through here, but it's getting late...
We might still need to reset `language` and `chapters` here.
If someone unloads an asset and then re-loads the same asset, if we don't reset those values on unload won't we think that the language and chapters did not actually change, and thus end up not making new chapter elements?

Reply:
If you are right, I had only tried changing the asset, but not loading it again. It is needed.